content/src/policies/rebac.check.rego

```diff
@@ -2,19 +2,38 @@ package rebac.check
 
 # default to a closed system (deny by default)
 default allowed = false
+default subject_type = "user"
 
 # resource context is expected in the following form:
 # {
 #   "relation": "relation or permission name",
 #   "object_type": "object type that carries the relation or permission",
 #   "object_id": "id of object instance with type of object_type"
+#   "subject_type": "[optional] subject type accessing the object. default is 'user'",
 # }
+#
+# To perform ReBAC checks with a subject that is not a user:
+# * set 'identity_context.type' to 'IDENTITY_TYPE_MANUAL'.
+# * set `identity_context.identity` to the subject ID.
+# * set `resource_context.subject_type` to the subject type.
 allowed {
   ds.check({
     "object_type": input.resource.object_type,
     "object_id": input.resource.object_id,
     "relation": input.resource.relation,
-    "subject_type": "user",
-    "subject_id": input.user.id
+    "subject_type": subject_type,
+    "subject_id": subject_id
   })
 }
+
+# When using IDENTITY_TYPE_MANUAL, the subject type comes from the resource context.
+subject_type := input.resource.subject_type if {
+    input.identity.type == "IDENTITY_TYPE_MANUAL"
+    input.resource.subject_type != ""
+}
+
+# When using IDENTITY_TYPE_MANUAL, the subject ID comes from the identity context.
+subject_id := input.identity.identity if {
```

Comment:
The logical conclusion of this could be that we also accept `subject_id` from `resource_context.subject_id` if the `input.identity.type` is `IDENTITY_TYPE_ANONYMOUS`.

Reply:
We discussed it yesterday but I'll reply here too for the record.

 I think that handling IDENTITY_TYPE_NONE in a ReBAC policy is tricky for several reasons:
* IDENTITY_TYPE_NONE is meant to indicate that the identity of the caller is unknown. But a ReBAC check _must_ include a subject. If one doesn't know who the subject is they can't make a `ds.check()` call in the first place so why are they using a ReBAC policy?
* Passing a subject_id in the resource context subverts the premise of IDENTITY_TYPE_NONE because clearly the caller does in fact know the identity.
* It also blurs the difference between IDENTITY_TYPE_NONE and IDENTITY_TYPE_MANUAL. We added IDENTITY_TYPE_MANUAL precisely for these kinds of use cases. We'd be creating two different ways to do the same thing:
  * Set the identity context to `{"type": IDENTITY_TYPE_MANUAL, "identity": "<subject id>"}`
  * Set the identity context to `{"type": IDENTITY_TYPE_NONE}` and send `"subject_id": "<subject id>"` in the resource context.
* What would be the expected behavior when someone uses IDENTITY_TYPE_MANUAL and sends one subject ID in the identity context and another in the resource context?